### DISCUSSION TITLE
blog/building-typescript-node-apps-with-nix

### DISCUSSION THREAD

**Discussion by giscus** [JRMurr/JRMurr.github.io#6] [Comments]
# blog/building-typescript-node-apps-with-nix

Trying some different nix builders for typescript node apps

https://johns.codes/blog/building-typescript-node-apps-with-nix

**Comment by GENDRAUD** [comment]
Hello,

First I would like to thank you for you post.
I've tried to do something similar.  I try to make development environment with a npm package that has not a nix equivalent.
I think that I'm very near from the end but I probably miss something important and it doesn't work.
I've edited a question on stackoverflow. https://stackoverflow.com/questions/75033884/how-to-modify-this-flake-nix-so-that-i-dont-have-to-load-the-nix-package-every
Your post has already helped me a lot but I need a little bit more help.
Would you mind telling me what is wrong in code?

**Reply by JRMurr** [reply]
Added an answer, let me know if this helps https://stackoverflow.com/questions/75033884/how-to-modify-this-flake-nix-so-that-i-dont-have-to-load-the-nix-package-every/75085113#75085113

**Reply by GENDRAUD** [reply]
I get the following error.

error: value is a function while a set was expected

devShell = npm_pack.shell;
                  ^

I'm surprised in default.nix 6 arguments are given to node-package.nix

import ./node-package.nix {
                         (1)             (2)              (3)        (4) (5)
   inherit (pkgs) fetchurl nix-gitignore stdenv lib fetchgit;
  inherit nodeEnv
                 (6)
}
 and node-packagea.nix has 6 compulsory arguments (the same as given) and a seventh argument not compulsory

{nodeEnv, fetchurl, fetchgit, nix-gitignore, stdenv, lib, globalBuildInputs ? []}:



npm_pack = import (./nix ); should be a function

**Reply by JRMurr** [reply]
O whoops my bad. The import should be `npm_pack = import ./nix { inherit pkgs system;};`

**Reply by GENDRAUD** [reply]
It seems to work! Thank you very much!
I don't accept the answer now because I only enter the environment successfully. I have not tested the package and I don't have time now. But as
soon as I can I accept the answer

**Comment by zestsystem** [comment]
Hi, Thanks for this writeup! The fastify example with Node2Nix works as a standalone index file but as soon as I use relative import in the index.ts file: 
```ts

#!/usr/bin/env node
import fastify from 'fastify';

import DbPlugin from './plugins/db';

const server = fastify();

server.register(DbPlugin);

server.get('/ping', async (request, reply) => {
    return 'pong\n';
});

server.listen({ host: '0.0.0.0', port: 8080 }, (err, address) => {
    if (err) {
        console.error(err);
        process.exit(1);
    }
    console.log(`Server listening at ${address}`);
});

```
I get this error:
`Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/nix/store/f2wvbx0jzng243w5izm8lqslkh03aa6g-marco/bin/plugins/db`

Does nix have a way to get around this?

**Reply by zestsystem** [reply]
Figured it out.
 In the install phase: 

```nix
          installPhase = ''
            runHook preInstall
            # Note: you need some sort of `mkdir` on $out for any of the following commands to work
            mkdir -p $out/bin
            # copy only whats needed for running the built app
            cp package.json $out/package.json
            cp -r dist $out/dist
            ln -sf ${nodeDeps}/lib/node_modules $out/node_modules

            # copy entry point, in this case our index.ts has the node shebang
            # nix will patch the shebang to be the node version specified in buildInputs
            # you could also copy in a script that is basically `npm run start`
            cp -r dist/* $out/bin // add other contents here
            cp -r dist/index.js $out/bin/app
            chmod a+x $out/bin/app
            runHook postInstall
          '';

```
